gpio/src/gpio.rs

```diff
@@ -0,0 +1,947 @@
+        Ok(match rtype {
+            VIRTIO_GPIO_MSG_GET_LINE_NAMES => self.gpio_names.as_bytes().to_vec(),
+            VIRTIO_GPIO_MSG_GET_DIRECTION => vec![self.get_direction(gpio)?],
+            VIRTIO_GPIO_MSG_SET_DIRECTION => {
+                self.set_direction(gpio, value)?;
+                vec![0]
+            }
+            VIRTIO_GPIO_MSG_GET_VALUE => vec![self.get_value(gpio)?],
+            VIRTIO_GPIO_MSG_SET_VALUE => {
+                self.set_value(gpio, value)?;
+                vec![0]
+            }
+            VIRTIO_GPIO_MSG_IRQ_TYPE => {
+                self.set_irq_type(gpio, value)?;
+                vec![0]
+            }
+            msg => return Err(Error::GpioMessageInvalid(msg)),
+        })
+    }
+}
+
+#[cfg(test)]
+pub(crate) mod tests {
+    use std::mem::size_of_val;
+
+    use super::Error;
+    use super::*;
+
+    #[derive(Debug)]
+    pub(crate) struct DummyDevice {
+        ngpio: u16,
+        pub(crate) gpio_names: Vec<String>,
+        line_state: RwLock<Vec<LineState>>,
+        get_num_gpios_result: Result<u16>,
+        get_gpio_name_result: Result<String>,
+        get_direction_result: Result<u8>,
+        set_direction_result: Result<()>,
+        get_value_result: Result<u8>,
+        set_value_result: Result<()>,
+        set_irq_type_result: Result<()>,
+        pub(crate) wait_for_irq_result: Result<()>,
+    }
+
+    impl DummyDevice {
+        pub(crate) fn new(ngpio: u16) -> Self {
+            Self {
+                ngpio,
+                gpio_names: vec!['\0'.to_string(); ngpio.into()],
+                line_state: RwLock::new(vec![
+                    LineState {
+                        dir: VIRTIO_GPIO_DIRECTION_NONE,
+                        val: None,
+                        irq_type: VIRTIO_GPIO_IRQ_TYPE_NONE,
+                    };
+                    ngpio.into()
+                ]),
+                get_num_gpios_result: Ok(0),
+                get_gpio_name_result: Ok("".to_string()),
+                get_direction_result: Ok(0),
+                set_direction_result: Ok(()),
+                get_value_result: Ok(0),
+                set_value_result: Ok(()),
+                set_irq_type_result: Ok(()),
+                wait_for_irq_result: Ok(()),
+            }
+        }
+    }
+
+    impl GpioDevice for DummyDevice {
+        fn open(_device: u32) -> Result<Self>
+        where
+            Self: Sized,
+        {
+            Ok(DummyDevice::new(8))
+        }
+
+        fn get_num_gpios(&self) -> Result<u16> {
+            if self.get_num_gpios_result.is_err() {
+                return self.get_num_gpios_result;
+            }
+
+            Ok(self.ngpio)
+        }
+
+        fn get_gpio_name(&self, gpio: u16) -> Result<String> {
+            assert!((gpio as usize) < self.gpio_names.len());
+
+            if self.get_gpio_name_result.is_err() {
+                return self.get_gpio_name_result.clone();
```

Comment:
The .clone() is probably not needed.

Reply:
It is required, else we get this:

move occurs because `self.get_gpio_name_result` has type `std::result::Result<std::string::String, gpio::Error>`, which does not implement the `Copy` trait

The problem is that the string is part of the structure, and we can't move it.